File: samples/Lua/TestLua/Resources/luaScript/TouchesTest/Ball.lua

require "extern"
require "luaScript/VisibleRect"
require "luaScript/TouchesTest/Paddle"

Ball = class("Ball", function(texture)
    return cc.Sprite:createWithTexture(texture)
end)

Ball.__index = Ball

Ball.m_velocity = cc.p(0,0)

local M_PI = 3.1415926

function Ball:radius()
    local  size = self:getTexture():getContentSize()
    return size.width/2
end

function Ball:move(delta)
    local getPosition = cc.p(self:getPosition())
    local position = cc.pMul(self.m_velocity, delta)
    self:setPosition( cc.pAdd(getPosition, position) )
    
    if (getPosition.x > VisibleRect:right().x - self:radius()) then
        self:setPosition( cc.p( VisibleRect:right().x - self:radius(), getPosition.y) )
        self.m_velocity.x = self.m_velocity.x * -1
    elseif (getPosition.x < VisibleRect:left().x + self:radius()) then
        self:setPosition( cc.p(VisibleRect:left().x + self:radius(), getPosition.y) )
        self.m_velocity.x = self.m_velocity.x * -1
    end
end

function Ball:collideWithPaddle(paddle)
    local paddleRect = paddle:rect()
    local paddleGetPosition = cc.p(paddle:getPosition())
    local selfGetPosition = cc.p(self:getPosition())

    paddleRect.x = paddleRect.x + paddleGetPosition.x
    paddleRect.y = paddleRect.y + paddleGetPosition.y

    local lowY  = cc.rectGetMinY(paddleRect) 
    local midY  = cc.rectGetMidY(paddleRect)
    local highY = cc.rectGetMaxY(paddleRect)

    local leftX  = cc.rectGetMinX(paddleRect)
    local rightX = cc.rectGetMaxX(paddleRect)

    if (selfGetPosition.x > leftX and selfGetPosition.x < rightX)  then

        local hit = false
        local angleOffset = 0.0 

        if (selfGetPosition.y > midY and selfGetPosition.y <= highY + self:radius()) then
            self:setPosition( cc.p(selfGetPosition.x, highY + self:radius()) )
            hit = true
            angleOffset = M_PI / 2
        elseif (selfGetPosition.y < midY and selfGetPosition.y >= lowY - self:radius()) then
            self:setPosition( cc.p(selfGetPosition.x, lowY - self:radius()) )
            hit = true
            angleOffset = -M_PI / 2
        end

        if (hit) then
            local hitAngle = cc.pToAngleSelf(cc.pSub(paddleGetPosition, paddleGetPosition)) + angleOffset

            local scalarVelocity = cc.pGetLength(self.m_velocity) * 1.05
            local velocityAngle = -cc.pToAngleSelf(self.m_velocity) + 0.5 * hitAngle

            self.m_velocity = cc.pMul(cc.pForAngle(velocityAngle), scalarVelocity)
        end
    end   

end

function Ball:setVelocity(velocity)
    self.m_velocity = velocity
end

function Ball:getVelocity()
    return self.m_velocity
end

function Ball.ballWithTexture(aTexture)

    if(aTexture == nil) then
        cclog("in ballWithTexture aTexture == nil")
    end

    local ball = Ball.new(aTexture)
    --ball:autorelease()
    return ball
end

File: tests/cpp-tests/Classes/TouchesTest/Ball.cpp
#include "Ball.h"
#include "Paddle.h"
#include "../VisibleRect.h"

USING_NS_CC;

Ball::Ball(void)
{
}

Ball::~Ball(void)
{
}

float Ball::radius()
{
    return getTexture()->getContentSize().width / 2;
}

Ball* Ball::ballWithTexture(Texture2D* aTexture)
{
    Ball* pBall = new (std::nothrow) Ball();
    pBall->initWithTexture(aTexture);
    pBall->autorelease();

    return pBall;
}

void Ball::move(float delta)
{
    this->setPosition(getPosition() + _velocity * delta);
    
    if (getPosition().x > VisibleRect::right().x - radius()) 
    {
        setPosition(VisibleRect::right().x - radius(), getPosition().y);
        _velocity.x *= -1;
    } 
    else if (getPosition().x < VisibleRect::left().x + radius()) 
    {
        setPosition(VisibleRect::left().x + radius(), getPosition().y);
        _velocity.x *= -1;
    }
}

void Ball::collideWithPaddle(Paddle* paddle)
{
    auto paddleRect = paddle->getRect();
    paddleRect.origin.x += paddle->getPosition().x;
    paddleRect.origin.y += paddle->getPosition().y;
    
    float lowY  = paddleRect.getMinY();
    float midY  = paddleRect.getMidY();
    float highY = paddleRect.getMaxY();
    
    float leftX  = paddleRect.getMinX();
    float rightX = paddleRect.getMaxX();
    
    if (getPosition().x > leftX && getPosition().x < rightX) {
    
        bool hit = false;
        float angleOffset = 0.0f; 
        
        if (getPosition().y > midY && getPosition().y <= highY + radius()) 
        {
            setPosition(getPosition().x, highY + radius());
            hit = true;
            angleOffset = (float)M_PI / 2;
        }
        else if (getPosition().y < midY && getPosition().y >= lowY - radius()) 
        {
            setPosition(getPosition().x, lowY - radius());
            hit = true;
            angleOffset = -(float)M_PI / 2;
        }
        
        if (hit) 
        {
            float hitAngle = (paddle->getPosition() - getPosition()).getAngle() + angleOffset;
            
            float scalarVelocity = _velocity.getLength() * 1.05f;
            float velocityAngle = -_velocity.getAngle() + 0.5f * hitAngle;
            
            _velocity = Vec2::forAngle(velocityAngle) * scalarVelocity;
        }
    }    
} 
